### ISSUE TITLE
error[E0117] provides misleading explanation.

### ISSUE THREAD

**Issue by caius-iulius** [rust-lang/rust#84859]
Given the following code: [rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=impl%20Drop%20for%20u32%7B%7D%0A%0Afn%20main()%7B%0A%20%20%20%20println!(%22Hey!%22)%0A%7D)
(This is straight-up copypasted from the error docs for E0117)

```rust
impl Drop for u32{};

fn main(){
    println!("Hey!")
}
```

The current output is, amongst other things:

```
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/main.rs:1:1
  |
1 | impl Drop for u32{}
  | ^^^^^^^^^^^^^^---
  | |             |
  | |             `u32` is not defined in the current crate
  | impl doesn't use only types from inside the current crate
  |
  = note: define and implement a trait or new type instead
```

The sentence "impl doesn't use only types from inside the current crate" isn't clear to me.
I'm not completely sure on how to change it, maybe you could repeat the error description.

My rationale is that in this case, given the arbitrary type, the presence or absence of it in the crate isn't relevant, since you can't redefine u32.
In my opinion the message should instead be focused on the location of the trait.

**Comment by Soveu** [comment]
> My rationale is that in this case, given the arbitrary type, the presence or absence of it in the crate isn't relevant, since you can't redefine u32.

Well, in this case yes, you can't redefine `u32` because it's a built-in type, but consider this
```rust
use some_crate::SomeType;
impl Drop for SomeType {}
```
`SomeType` is implemented outside of your crate, but you can also define your own type named `SomeType`

The reason it's not possible in Rust to implement external traits for external types is that there can be other crates relying on `SomeType` not having certain traits (for example `Drop` or `Copy`) and it would just get messy with more and more crates.

It works in any other combination though:

1. internal trait for internal type
2. external trait for internal type
3. internal trait for external type

Thus, the error messages
```
only traits defined in the current crate can be implemented for arbitrary types
```
```
`u32` is not defined in the current crate
```
and a suggestion
```
note: define and implement a trait or new type instead
```

**Comment by spydx** [comment]
I would agree that the `err.span_label` for E0117 is a bad sentence : "impl doesn't use only types from inside the current crate".  Since I'm not a native english speaker this "doesn't use only" is what tosses me of the rails. 
The error message for E0117 is very good and explains it. 

The `err.span_label` could have become something like this e.g "impl can only use types from inside the current crate" to reflect the error message better.

**Comment by TheTechRobo** [comment]
Either that, or "impl uses types defined outside of the current crate" or something similar. The current sentence a bit hard to read at first.

**Comment by estebank** [comment]
Current output:

```
error[E0117]: only traits defined in the current crate can be implemented for primitive types
 --> src/main.rs:1:1
  |
1 | impl Drop for u32{}
  | ^^^^^^^^^^^^^^---
  |               |
  |               `u32` is not defined in the current crate
  |
  = note: impl doesn't have any local type before any uncovered type parameters
  = note: for more information see https://doc.rust-lang.org/reference/items/implementations.html#orphan-rules
  = note: define and implement a trait or new type instead

error[E0120]: the `Drop` trait may only be implemented for local structs, enums, and unions
 --> src/main.rs:1:15
  |
1 | impl Drop for u32{}
  |               ^^^ must be a struct, enum, or union in the current crate
```
